### PR TITLE
Backport loadwallet

### PR DESCRIPTION
**Pull request by PastaPastaPasta**
Backports 13028, 10740, 11851.
(see comment edit log to view seg fault that was occurring that has now been resolved)

### CODE REVIEW

**Comment by xdustinface** [comment]
> and when using `loadwallet` from QT, a segmentation fault occurs

This helps here https://github.com/xdustinface/dash/commit/303df94f794f3ac668d9edb77681af6c7e979dd0

File: src/wallet/rpcwallet.cpp [hunk 1]
```diff
@@ -2880,6 +2880,53 @@ UniValue keepass(const JSONRPCRequest& request)
     return "Invalid command";
 }
 
+UniValue loadwallet(const JSONRPCRequest& request)
+{
+    if (request.fHelp || request.params.size() != 1)
+        throw std::runtime_error(
+            "loadwallet \"filename\"\n"
+            "\nLoads a wallet from a wallet file or directory."
+            "\nNote that all wallet command-line options used when starting bitcoind will be"
```
Comment (xdustinface) [inline src/wallet/rpcwallet.cpp:2889]:
```suggestion
            "\nNote that all wallet command-line options used when starting dashd will be"
```

**Comment by UdjinM6** [comment]
This requires many more backports for correct handling `-nofoo` cmd-line args/config options because of `-nowallet` - it doesn't work as expected atm.

**Comment by PastaPastaPasta** [comment]
@UdjinM6 Can you point me to those PRs if you have any idea?

**Comment by UdjinM6** [comment]
Pls see https://github.com/UdjinM6/dash/commits/pr3462. It's really #12713 + 4d34fcc7138f0ffc831f0f8601c50cc7f494c197 from #11862 that are needed but had to backport other PRs to avoid (large) merge conflicts. Might backport a couple of first commits instead of a full 11862 if network specific config sections is not smth we want (right now). Pls note, I did NOT yet throughly reviewed/tested these, just made sure changes somewhat made sense when resolving (trivial-ish) merge conflicts and that it compiles and integration tests pass locally.

**Comment by PastaPastaPasta** [comment]
backported, have not tested yet. Hopefully will test tonight. Looks like scripted diff fails. Usual

**Comment by PastaPastaPasta** [comment]
Mostly seems to work, but when using loadwallet in qt results in core dump

```
dash-qt: wallet/db.cpp:233: BerkeleyEnvironment::VerifyResult BerkeleyEnvironment::Verify(const string&, BerkeleyEnvironment::recoverFunc_type, std::__cxx11::string&): Assertion `mapFileUseCount.count(strFile) == 0' failed.
Posix Signal: Aborted
   0#: (0x557BC454A135) stl_vector.h:466    - std::vector<unsigned long, std::allocator<unsigned long> >::operator=(std::vector<unsigned long, std::allocator<unsigned long> >&&)
   1#: (0x557BC454A135) stacktraces.cpp:779 - HandlePosixSignal
   2#: (0x7FAD26A30890) <unknown-file>      - ???
   3#: (0x7FAD25B28E97) <unknown-file>      - ???
   4#: (0x7FAD25B2A801) <unknown-file>      - ???
   5#: (0x7FAD25B1A39A) <unknown-file>      - ???
   6#: (0x7FAD25B1A412) <unknown-file>      - ???
   7#: (0x557BC442F3A8) mutex:111           - std::recursive_mutex::lock()
   8#: (0x557BC442F3A8) sync.h:59           - AnnotatedMixin<std::recursive_mutex>::lock()
   9#: (0x557BC442F3A8) std_mutex.h:267     - std::unique_lock<CCriticalSection>::lock()
  10#: (0x557BC442F3A8) sync.h:128          - CCriticalBlock::Enter(char const*, char const*, int)
  11#: (0x557BC442F3A8) sync.h:149          - CCriticalBlock::CCriticalBlock(CCriticalSection&, char const*, char const*, int, bool)
  12#: (0x557BC442F3A8) db.cpp:232          - BerkeleyEnvironment::Verify(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool (*)(boost::filesystem::path const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&), std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&)
  13#: (0x557BC442F61D) db.cpp:350          - BerkeleyBatch::VerifyDatabaseFile(boost::filesystem::path const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, bool (*)(boost::filesystem::path const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&))
  14#: (0x557BC43D43FC) wallet.cpp:5062     - CWallet::Verify(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, bool, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&)
  15#: (0x557BC4395F90) basic_string.h:211  - std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_is_local() const
  16#: (0x557BC4395F90) basic_string.h:220  - std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_dispose()
  17#: (0x557BC4395F90) basic_string.h:647  - std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::~basic_string()
  18#: (0x557BC4395F90) rpcwallet.cpp:2911  - loadwallet(JSONRPCRequest const&)
  19#: (0x557BC41BA26D) server.cpp:565      - CRPCTable::execute(JSONRPCRequest const&) const
  20#: (0x557BC3E10F0A) univalue.h:20       - UniValue::operator=(UniValue&&)
  21#: (0x557BC3E10F0A) rpcconsole.cpp:323  - RPCConsole::RPCParseCommandLine(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*)
  22#: (0x557BC3E117A3) rpcconsole.cpp:429  - RPCExecutor::request(QString const&)
  23#: (0x557BC3E11F69) rpcconsole.moc:78   - RPCExecutor::qt_static_metacall(QObject*, QMetaObject::Call, int, void**)
  24#: (0x557BC5169FF2) <unknown-file>      - ???
Aborted (core dumped)
```

**Comment by xdustinface** [comment]
> Mostly seems to work, but when using loadwallet in qt results in core dump

c7e6adf86963c62a146857d8958cc9b943d3bbd0  This one fixes the crash.. Happened because the wallet object had no name set by default. This led to not keeping track of `wallet.dat` as used file and then to this failing assertion `assert(mapFileUseCount.count(strFile) == 0)`.

Also feel free to apply those two if you want

1. 9d811e8e39d5621b2de1b3d58925c615519cdfb9 Use a constant for the default name in all other places of the code

2. e412f598def029b213af23a42a368b1882c6d139  and to fully avoid this crash throw an error if someone starts with `-wallet=""`

**Comment by PastaPastaPasta** [comment]
According to @xdustinface, the crash that I pointed out actually occurs in bitcoin too, maybe this is an okay crash to have rn? Thoughts @UdjinM6

**Comment by UdjinM6** [comment]
Going back to default _file_ name is not a good idea imo, because `-wallet` was redefined to specify a path, not a file. But crashing is not good either, would prefer at least some meaningful response back to the user, smth like e412f59 proposed above but maybe saying "-wallet can't be an empty string" or smth like that (i.e. mentioning the exact setting that caused an error and not mentioning "file" cause it's not a file anymore).

**Comment by PastaPastaPasta** [comment]
@UdjinM6 @xdustinface Regarding this particular crash, I think it's okay for now, bitcoin had it https://github.com/bitcoin/bitcoin/issues/14538 fixed by https://github.com/bitcoin/bitcoin/pull/14552 however 14552 is locked behind some further backporting. After this is merged I might be able to backport the required stuff to fix this bug in 0.16, if we don't have enough time we just note it as a known issue in release notes.

**Comment by PastaPastaPasta** [comment]
force pushed doing two things, one, based this branch on #3472, two fixed scripted diff

**Comment by PastaPastaPasta** [comment]
In testing, everything seems to work, I'm able to load a directory based wallet, and I'm able to load a file based wallet. Everything seems dandy to me. 

Ofc there is that one crash when you are using wallet="" and try to load wallet.dat, but as I pointed out above, I think that is fine for now.

**Comment by PastaPastaPasta** [comment]
squashed all the 10740 changes into one commit to make this PR nicer

File: doc/release-notes-pr12823.md [hunk 2]
```diff
@@ -0,0 +1,20 @@
+Configuration sections for testnet and regtest
```
Comment (UdjinM6) [inline doc/release-notes-pr12823.md:1]: Should either keep one from 10740 or drop this file too for consistency sake.

File: src/wallet/init.cpp [hunk 3]
```diff
@@ -253,63 +253,29 @@ bool WalletInit::Verify()
...
-            return InitError(strError);
-        }
-
-        std::string strWarning;
-        std::unique_ptr<CWallet> tempWallet = MakeUnique<CWallet>(walletFile, CWalletDBWrapper::Create(wallet_path));
-        if (!tempWallet->AutoBackupWallet(wallet_path, strWarning, strError)) {
-            if (!strWarning.empty())
-                InitWarning(strWarning);
-            if (!strError.empty())
-                return InitError(strError);
```
Comment (UdjinM6) [inline src/wallet/init.cpp:293]: "Aaaaand it's gone"? Pls see be1253e996

**Comment by PastaPastaPasta** [comment]
see latest, removed release-note doc and squashed, and cherry-picked be1253e996ae44dd56e437a6fdf3d8abbbeced8f

File: src/wallet/rpcwallet.cpp [hunk 4]
```diff
@@ -3450,6 +3497,7 @@ static const CRPCCommand commands[] =
     { "wallet",             "listtransactions",         &listtransactions,         {"account","count","skip","include_watchonly"} },
     { "wallet",             "listunspent",              &listunspent,              {"minconf","maxconf","addresses","include_unsafe","query_options"} },
     { "wallet",             "listwallets",              &listwallets,              {} },
+    { "wallet",             "loadwallet",                       &loadwallet,                    {"filename"} },
```
[the remark this answers is not in the document]
Comment (PastaPastaPasta) [reply src/wallet/rpcwallet.cpp]: committed, and then squashed

**Pull request merged**